components/usage/pkg/apiv1/billing_test.go

```diff
@@ -125,3 +125,190 @@ func TestFinalizeInvoice(t *testing.T) {
 
 	require.NoError(t, err)
 }
+
+func TestCapUsageToSpendingLimit(t *testing.T) {
+	type ReportSummary struct {
+		// These are the inputs
+		CalculatedUsage int32
+		SpendingLimit   int32
+		// These are the outputs (also depending on the previous run)
+		EffectiveSpendingLimit int32
+		CappedUsage            int32
+	}
+
+	cap := func(v, limit int32) int32 {
+		if v > limit {
+			return limit
+		}
+		return v
+	}
+
+	max := func(a, b int32) int32 {
+		if a >= b {
+			return a
+		} else {
+			return b
+		}
+	}
+
+	min := func(a, b int32) int32 {
+		if a <= b {
+			return a
+		} else {
+			return b
+		}
+	}
+
+	/**
+	 * Takes two ReportSummaries, both from the _same invoice_ (!).
+	 */
+	capUsage := func(previous, current ReportSummary) (cappedUsage int32, effectiveLimit int32) {
+		// idea: if a user reduces spending limit during a billing cycle, we should:
+		//  - allow to reduce that threshold
+		//  - but not below what we currently plan to bill them
```

Comment:
I made the assumption that `EffectiveSpendingLimit` in these test cases is referring to "what we currently plan to bill them" aka `upcomingInvoice`. Is that correct?

Reply:
@laushinka No, that's `cappedUsage`. Sorry for not making that clearer: It's a value that we retrieve from the metadata of the `upcomingInvoide`, prior to actually updating it. It's a value we carry over (and manipulate) with every run.